File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/service/AudioNewsService.java
package com.andruid.magic.newsdaily.service;

import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Message;
import android.support.v4.media.MediaBrowserCompat.MediaItem;
import android.support.v4.media.MediaDescriptionCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.media.MediaBrowserServiceCompat;
import androidx.media.session.MediaButtonReceiver;
import androidx.preference.PreferenceManager;

import com.andruid.magic.newsdaily.R;
import com.andruid.magic.newsdaily.model.AudioNews;
import com.andruid.magic.newsdaily.util.Extensions;
import com.andruid.magic.newsdaily.util.NotificationUtil;
import com.andruid.magic.newsdaily.util.PrefUtil;
import com.andruid.magic.newsloader.api.NewsRepository;
import com.andruid.magic.newsloader.model.News;
import com.andruid.magic.texttoaudiofile.api.TtsApi;
import com.andruid.magic.texttoaudiofile.util.FileUtils;
import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.audio.AudioAttributes;
import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector;
import com.google.android.exoplayer2.ext.mediasession.TimelineQueueNavigator;
import com.google.android.exoplayer2.source.ConcatenatingMediaSource;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.TrackSelector;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.util.Util;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import timber.log.Timber;

import static com.andruid.magic.newsdaily.data.Constants.ACTION_PREPARE_AUDIO;
import static com.andruid.magic.newsdaily.data.Constants.EXTRA_CATEGORY;
import static com.andruid.magic.newsloader.data.Constants.FIRST_PAGE;
import static com.andruid.magic.newsloader.data.Constants.PAGE_SIZE;

public class AudioNewsService extends MediaBrowserServiceCompat implements Player.EventListener,
        NewsRepository.NewsLoadedListener, TtsApi.AudioConversionListener {
    private static final String MEDIA_SERVICE = "AudioNewsService";
    public static final int NEWS_FETCH_DISTANCE = 3, MEDIA_NOTI_ID = 1;
    private static final int MSG_INIT_NEWS = 0, MSG_STOP_SERVICE = 1, MSG_UPDATE_SOURCE = 2, MSG_SHOW_NOTI = 3;
    private static final int WAIT_QUEUE_TIMEOUT_MS = 5000;
    private MediaSessionCompat mediaSessionCompat;
    private MediaSessionCallback mediaSessionCallback;
    private SimpleExoPlayer exoPlayer;
    private MediaSessionConnector mediaSessionConnector;
    private List<AudioNews> audioNewsList = new ArrayList<>();
    private DataSource.Factory dataSourceFactory;
    private ConcatenatingMediaSource concatenatingMediaSource;
    private Intent mediaButtonIntent;
    private int page = FIRST_PAGE;
    private BroadcastReceiver mNoisyReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            mediaSessionCallback.onPause();
        }
    };
    private String category;
    private Handler audioNewsHandler;
    private HandlerThread audioNewsHandlerThread;

    @Override
    public void onCreate() {
        super.onCreate();
        Timber.d("service created");
        audioNewsHandlerThread = new HandlerThread("HandlerThread");
        audioNewsHandlerThread.start();
        audioNewsHandler = new Handler(audioNewsHandlerThread.getLooper(), new AudioNewsHandler());
        initMediaSession();
        initExoPlayer();
    }

    private void initExoPlayer() {
        final TrackSelector trackSelector = new DefaultTrackSelector();
        exoPlayer = ExoPlayerFactory.newSimpleInstance(this, trackSelector);
        dataSourceFactory = new DefaultDataSourceFactory(getApplicationContext(),
                Util.getUserAgent(this, getString(R.string.app_name)));
        AudioAttributes audioAttributes = new AudioAttributes.Builder()
                .setUsage(C.USAGE_MEDIA)
                .setContentType(C.CONTENT_TYPE_MUSIC)
                .build();
        exoPlayer.setAudioAttributes(audioAttributes,true);
        exoPlayer.addListener(this);
        concatenatingMediaSource = new ConcatenatingMediaSource();
    }

    private void initMediaSession() {
        Intent mediaButtonIntent = new Intent(Intent.ACTION_MEDIA_BUTTON);
        mediaButtonIntent.setClass(this, MediaButtonReceiver.class);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(this, 0,
                mediaButtonIntent, 0);
        ComponentName mediaButtonReceiver = new ComponentName(getApplicationContext(),
                MediaButtonReceiver.class);
        mediaSessionCompat = new MediaSessionCompat(getApplicationContext(), MEDIA_SERVICE,
                mediaButtonReceiver, pendingIntent);
        setSessionToken(mediaSessionCompat.getSessionToken());
        mediaSessionCallback = new MediaSessionCallback();
        mediaSessionCompat.setCallback(mediaSessionCallback);
        mediaSessionCompat.setActive(true);
        mediaSessionConnector = new MediaSessionConnector(mediaSessionCompat);
        mediaSessionConnector.setQueueNavigator(new TimelineQueueNavigator(mediaSessionCompat) {
            @Override
            public MediaDescriptionCompat getMediaDescription(Player player, int windowIndex) {
                return Extensions.getMediaDescription(audioNewsList.get(windowIndex));
            }
        });
        mediaSessionConnector.setPlayer(exoPlayer,null);
    }

    private void loadNews(){
        String country = PreferenceManager.getDefaultSharedPreferences(this).getString(
                getString(R.string.pref_country), PrefUtil.getDefaultCountry(this));
        NewsRepository.getInstance().loadHeadlines(country, category, page, PAGE_SIZE, this);
    }

    private void setCurrentAudio(int pos) {
        if(pos >= audioNewsList.size() || pos < 0)
            return;
        AudioNews audioNews = audioNewsList.get(pos);
        mediaSessionCompat.setMetadata(Extensions.buildMetaData(audioNews));
    }

    private void setMediaPlaybackState(int state, int pos) {
        PlaybackStateCompat.Builder playbackStateBuilder = new PlaybackStateCompat.Builder();
        playbackStateBuilder.setActiveQueueItemId(pos);
        if (state == PlaybackStateCompat.STATE_PLAYING || state == PlaybackStateCompat.STATE_BUFFERING ||
                state == PlaybackStateCompat.STATE_SKIPPING_TO_PREVIOUS || state == PlaybackStateCompat.STATE_SKIPPING_TO_NEXT)
            playbackStateBuilder.setActions(PlaybackStateCompat.ACTION_PLAY_PAUSE | PlaybackStateCompat.ACTION_PAUSE | PlaybackStateCompat.ACTION_SKIP_TO_NEXT | PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS | PlaybackStateCompat.ACTION_STOP);
        else
            playbackStateBuilder.setActions(PlaybackStateCompat.ACTION_PLAY_PAUSE | PlaybackStateCompat.ACTION_PLAY | PlaybackStateCompat.ACTION_SKIP_TO_NEXT | PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS | PlaybackStateCompat.ACTION_STOP);
        if (state == PlaybackStateCompat.STATE_PAUSED || state == PlaybackStateCompat.STATE_PLAYING) {
            try {
                playbackStateBuilder.setState(state, exoPlayer.getCurrentPosition(), 0);
            }
            catch (IllegalStateException e){
                e.printStackTrace();
                Toast.makeText(getApplicationContext(),e.getMessage(),Toast.LENGTH_SHORT).show();
            }
        } else
            playbackStateBuilder.setState(state, PlaybackStateCompat.PLAYBACK_POSITION_UNKNOWN, 0);
        mediaSessionCompat.setPlaybackState(playbackStateBuilder.build());
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if(ACTION_PREPARE_AUDIO.equals(intent.getAction())) {
            Bundle extras = intent.getExtras();
            if(extras != null)
                category = extras.getString(EXTRA_CATEGORY);
            if(TtsApi.getInstance().isReady())
                audioNewsHandler.sendEmptyMessage(MSG_INIT_NEWS);
            else
                Toast.makeText(getApplicationContext(), "Text to speech not available", Toast.LENGTH_SHORT).show();
            registerReceiver(mNoisyReceiver, new IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY));
        }
        else
            MediaButtonReceiver.handleIntent(mediaSessionCompat, intent);
        return START_NOT_STICKY;
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        super.onTaskRemoved(rootIntent);
        audioNewsHandler.sendEmptyMessage(MSG_STOP_SERVICE);
    }

    @Override
    public void onDestroy() {
        mediaSessionCompat.release();
        mediaSessionConnector.setPlayer(null, null);
        try {
            unregisterReceiver(mNoisyReceiver);
        } catch (IllegalArgumentException e) {
            e.printStackTrace();
        }
        exoPlayer.release();
        audioNewsHandlerThread.quit();
    }

    @Override
    public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
        if(playbackState == Player.STATE_READY) {
            if (mediaButtonIntent != null) {
                MediaButtonReceiver.handleIntent(mediaSessionCompat, mediaButtonIntent);
                mediaButtonIntent = null;
            }
            int icon = android.R.drawable.ic_media_play;
            if (playWhenReady)
                icon = android.R.drawable.ic_media_pause;
            MediaMetadataCompat metadataCompat = mediaSessionCompat.getController().getMetadata();
            if (metadataCompat == null || category == null)
                return;
            Timber.d("sending message noti show in state change");
            Message message = audioNewsHandler.obtainMessage(MSG_SHOW_NOTI, icon, playWhenReady ? 1 : 0, category);
            audioNewsHandler.sendMessage(message);
        }
    }

    @Override
    public void onPositionDiscontinuity(int reason) {
        if(reason==Player.DISCONTINUITY_REASON_SEEK_ADJUSTMENT ||
                reason==Player.DISCONTINUITY_REASON_PERIOD_TRANSITION || reason==Player.DISCONTINUITY_REASON_INTERNAL){
            int pos = exoPlayer.getCurrentWindowIndex();
            if(pos == audioNewsList.size() - NEWS_FETCH_DISTANCE) {
                page++;
                loadNews();
            }
            AudioNews audioNews = audioNewsList.get(pos);
            MediaMetadataCompat mediaMetadataCompat = Extensions.buildMetaData(audioNews);
            mediaSessionCompat.setMetadata(mediaMetadataCompat);
            Timber.d("sending message noti show in pos discontinuity");
            Message message = audioNewsHandler.obtainMessage(MSG_SHOW_NOTI, android.R.drawable.ic_media_pause,
                    1, category);
            audioNewsHandler.sendMessage(message);
        }
    }

    @Override
    public void onSuccess(List<News> newsList, boolean hasMore) {
        int pos = audioNewsList.size();
        for(int i=0; i<newsList.size(); i++){
            News news = newsList.get(i);
            String text = news.getDesc();
            audioNewsList.add(new AudioNews(news.getUrl(), news));
            if(text.isEmpty())
                text = "No description available";
            TtsApi.getInstance().convertToAudioFile(text, String.valueOf(pos+i), this);
        }
    }

    @Override
    public void onFailure(Throwable t) {
        t.printStackTrace();
    }

    @Override
    public BrowserRoot onGetRoot(@NonNull String clientPackageName, int clientUid, Bundle rootHints) {
        return new BrowserRoot("root", null);
    }

    @Override
    public void onLoadChildren(@NonNull final String parentMediaId,
                               @NonNull final Result<List<MediaItem>> result) {
        result.sendResult(new ArrayList<>());
    }

    @Override
    public void onAudioCreated(File file) {
        String utteranceId = FileUtils.getUtteranceId(file.getName());
        Timber.d("Utterance done %s", utteranceId);
        MediaSource mediaSource = new ExtractorMediaSource.Factory(dataSourceFactory)
                .createMediaSource(Uri.fromFile(file));
        concatenatingMediaSource.addMediaSource(mediaSource);
        audioNewsHandler.sendEmptyMessageDelayed(MSG_UPDATE_SOURCE, WAIT_QUEUE_TIMEOUT_MS);
        if(utteranceId.equals("0")) {
            audioNewsHandler.sendEmptyMessage(MSG_UPDATE_SOURCE);
            mediaSessionCallback.onPlay();
        }
    }

    @Override
    public void onFailure(@NotNull String msg) {
        Toast.makeText(getApplicationContext(), msg, Toast.LENGTH_SHORT).show();
    }

    private class MediaSessionCallback extends MediaSessionCompat.Callback {
        @Override
        public void onPlay() {
            super.onPlay();
            if(audioNewsList == null)
                return;
            exoPlayer.setPlayWhenReady(true);
            int pos = exoPlayer.getCurrentWindowIndex();
            setCurrentAudio(pos);
            mediaSessionCompat.setActive(true);
            setMediaPlaybackState(PlaybackStateCompat.STATE_PLAYING, pos);
        }

        @Override
        public void onPause() {
            super.onPause();
            exoPlayer.setPlayWhenReady(false);
            setMediaPlaybackState(PlaybackStateCompat.STATE_PAUSED, exoPlayer.getCurrentWindowIndex());
        }

        @Override
        public void onSkipToNext() {
            super.onSkipToNext();
            if(exoPlayer.hasNext())
                exoPlayer.next();
        }

        @Override
        public void onSkipToPrevious() {
            super.onSkipToPrevious();
            if(exoPlayer.hasPrevious())
                exoPlayer.previous();
        }
    }

    private class AudioNewsHandler implements Handler.Callback {

        @Override
        public boolean handleMessage(@NonNull Message message) {
            Timber.d("handle message: %d", message.what);
            switch (message.what){
                case MSG_INIT_NEWS:
                    audioNewsHandler.removeCallbacksAndMessages(null);
                    audioNewsList.clear();
                    concatenatingMediaSource.clear();
                    page = FIRST_PAGE;
                    loadNews();
                    break;
                case MSG_STOP_SERVICE:
                    audioNewsHandler.removeCallbacksAndMessages(null);
                    stopForeground(true);
                    stopSelf();
                    break;
                case MSG_UPDATE_SOURCE:
                    audioNewsHandler.removeMessages(MSG_UPDATE_SOURCE);
                    exoPlayer.prepare(concatenatingMediaSource, false, false);
                    break;
                case MSG_SHOW_NOTI:
                    int icon = message.arg1;
                    String category = (String) message.obj;
                    boolean playWhenReady = (message.arg2 == 1);
                    MediaMetadataCompat metadataCompat = mediaSessionCompat.getController().getMetadata();
                    NotificationCompat.Builder notificationBuilder = NotificationUtil.buildNotification(AudioNewsService.this,
                            icon, category, metadataCompat, mediaSessionCompat.getSessionToken());
                    if(playWhenReady)
                        startForeground(MEDIA_NOTI_ID, notificationBuilder.build());
                    else {
                        NotificationManagerCompat.from(AudioNewsService.this).notify(MEDIA_NOTI_ID,
                                notificationBuilder.build());
                        stopForeground(false);
                    }
                    break;
            }
            return false;
        }
    }
}

File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/service/AudioNewsService.kt
package com.andruid.magic.newsdaily.service

import android.app.NotificationManager
import android.app.PendingIntent
import android.content.ComponentName
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.support.v4.media.MediaBrowserCompat.MediaItem
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.util.Log
import androidx.core.content.getSystemService
import androidx.media.MediaBrowserServiceCompat
import androidx.media.session.MediaButtonReceiver
import com.andruid.magic.newsdaily.R
import com.andruid.magic.newsdaily.data.ACTION_PREPARE_AUDIO
import com.andruid.magic.newsdaily.data.EXTRA_CATEGORY
import com.andruid.magic.newsdaily.data.PAGE_SIZE
import com.andruid.magic.newsdaily.data.model.AudioNews
import com.andruid.magic.newsdaily.data.model.buildMetaData
import com.andruid.magic.newsdaily.data.model.getMediaDescription
import com.andruid.magic.newsdaily.data.model.toAudioNews
import com.andruid.magic.newsdaily.database.repository.DbRepository
import com.andruid.magic.newsdaily.util.buildNotification
import com.andruid.magic.newsdaily.util.getDefaultCountry
import com.andruid.magic.newsdaily.util.toast
import com.andruid.magic.texttoaudiofile.api.TtsApi
import com.andruid.magic.texttoaudiofile.data.model.TtsResult
import com.andruid.magic.texttoaudiofile.util.FileUtils.getUtteranceId
import com.google.android.exoplayer2.C
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.audio.AudioAttributes
import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector
import com.google.android.exoplayer2.ext.mediasession.TimelineQueueNavigator
import com.google.android.exoplayer2.source.ConcatenatingMediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.google.android.exoplayer2.util.Util
import kotlinx.coroutines.*
import java.util.*
import kotlin.coroutines.CoroutineContext

class AudioNewsService : MediaBrowserServiceCompat(), CoroutineScope, Player.EventListener {
    companion object {
        private val TAG = "${AudioNewsService::class.java.simpleName}Log"

        private const val MEDIA_SERVICE = "AudioNewsService"
        private const val MSG_INIT_NEWS = 0
        private const val MSG_STOP_SERVICE = 1
        private const val MSG_UPDATE_SOURCE = 2
        private const val MSG_SHOW_NOTI = 3
        private const val WAIT_QUEUE_TIMEOUT_MS = 5000

        private const val NEWS_FETCH_DISTANCE = 3
        private const val MEDIA_NOTI_ID = 1
    }

    private val job = Job()
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job

    private val mediaSessionCompat by lazy {
        val mediaButtonIntent = Intent(Intent.ACTION_MEDIA_BUTTON).setClass(
            this,
            MediaButtonReceiver::class.java
        )
        val pendingIntent = PendingIntent.getBroadcast(this, 0, mediaButtonIntent, 0)
        val mediaButtonReceiver = ComponentName(applicationContext, MediaButtonReceiver::class.java)
        MediaSessionCompat(
            applicationContext,
            MEDIA_SERVICE, mediaButtonReceiver, pendingIntent
        )
    }
    private val exoPlayer by lazy {
        SimpleExoPlayer.Builder(this)
            .build()
    }
    private val dataSourceFactory by lazy {
        DefaultDataSourceFactory(
            applicationContext,
            Util.getUserAgent(this, getString(R.string.app_name))
        )
    }
    private val mediaSessionConnector by lazy { MediaSessionConnector(mediaSessionCompat) }
    private val audioNewsHandler = Handler(AudioNewsHandler())
    private val concatenatingMediaSource = ConcatenatingMediaSource()
    private val mediaSessionCallback = MediaSessionCallback()
    private val audioNewsList = mutableListOf<AudioNews>()

    private var page = 0

    private lateinit var category: String

    override fun onCreate() {
        super.onCreate()
        Log.d(TAG, "service created")

        job.start()
        initExoPlayer()
        initMediaSession()
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        if (ACTION_PREPARE_AUDIO == intent.action) {
            category = intent.extras?.getString(EXTRA_CATEGORY) ?: "general"
            if (TtsApi.isReady)
                audioNewsHandler.sendEmptyMessage(MSG_INIT_NEWS)
            else
                toast("Text to speech not available")
        } else
            MediaButtonReceiver.handleIntent(mediaSessionCompat, intent)
        return START_NOT_STICKY
    }

    override fun onTaskRemoved(rootIntent: Intent?) {
        super.onTaskRemoved(rootIntent)
        audioNewsHandler.sendEmptyMessage(MSG_STOP_SERVICE)
    }

    override fun onDestroy() {
        job.cancel()
        mediaSessionCompat.release()
        mediaSessionConnector.setPlayer(null)
        exoPlayer.release()
    }

    override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
        if (playbackState == Player.STATE_READY) {
            if (mediaSessionCompat.controller.metadata == null)
                return
            val icon =
                if (playWhenReady) android.R.drawable.ic_media_pause else android.R.drawable.ic_media_play

            val message = audioNewsHandler.obtainMessage(
                MSG_SHOW_NOTI, icon,
                if (playWhenReady) 1 else 0,
                category
            )
            audioNewsHandler.sendMessage(message).also {
                Log.d(TAG, "sending message noti show in state change")
            }
        }
    }

    override fun onPositionDiscontinuity(reason: Int) {
        if (reason == Player.DISCONTINUITY_REASON_SEEK_ADJUSTMENT ||
            reason == Player.DISCONTINUITY_REASON_PERIOD_TRANSITION ||
            reason == Player.DISCONTINUITY_REASON_INTERNAL
        ) {
            val pos = exoPlayer.currentWindowIndex
            if (pos == audioNewsList.size - NEWS_FETCH_DISTANCE) {
                launch {
                    Log.d("audioLog", "reached news fetch distance pos = $pos")
                    loadNews()
                }
            }
            val audioNews = audioNewsList[pos]
            val mediaMetadataCompat = audioNews.buildMetaData()
            mediaSessionCompat.setMetadata(mediaMetadataCompat)

            val message = audioNewsHandler.obtainMessage(
                MSG_SHOW_NOTI,
                android.R.drawable.ic_media_pause,
                1,
                category
            )
            audioNewsHandler.sendMessage(message).also {
                Log.d(TAG, "sending message noti show in pos discontinuity")
            }
        }
    }

    override fun onGetRoot(clientPackageName: String, clientUid: Int, rootHints: Bundle?) =
        BrowserRoot("root", null)

    override fun onLoadChildren(parentMediaId: String, result: Result<List<MediaItem?>?>) {
        result.sendResult(ArrayList())
    }

    private suspend fun loadNews() {
        Log.d("audioLog", "load news called page =  $page, page size = $PAGE_SIZE")
        withContext(Dispatchers.IO) {
            DbRepository.getNewsForPage(
                getDefaultCountry(),
                category,
                page++,
                PAGE_SIZE
            )
        }
            .map { news -> news.toAudioNews() }
            .forEach { news ->
                val pos = audioNewsList.size
                Log.d("audioLog", "news = ${news.news.title}")
                val text =
                    if (news.news.desc?.isNotEmpty() == true) news.news.desc else "No description available"

                audioNewsList.add(news)
                addAudioToQueue(text, pos.toString())
            }
    }

    private suspend fun addAudioToQueue(text: String, id: String) {
        val result = TtsApi.convertToAudioFile(text, id)

        if (result is TtsResult.Success) {
            val file = result.file
            val utteranceId = getUtteranceId(file.name).also {
                Log.d(TAG, "Utterance done $it")
            }

            val mediaSource = ProgressiveMediaSource.Factory(dataSourceFactory)
                .createMediaSource(Uri.fromFile(file))
            concatenatingMediaSource.addMediaSource(mediaSource)

            if (utteranceId == "0") {
                audioNewsHandler.sendEmptyMessage(MSG_UPDATE_SOURCE)
                mediaSessionCallback.onPlay()
            } else {
                audioNewsHandler.sendEmptyMessageDelayed(
                    MSG_UPDATE_SOURCE,
                    WAIT_QUEUE_TIMEOUT_MS.toLong()
                )
            }
        }
    }

    private fun initExoPlayer() {
        exoPlayer.apply {
            val audioAttributes = AudioAttributes.Builder()
                .setUsage(C.USAGE_MEDIA)
                .setContentType(C.CONTENT_TYPE_MUSIC)
                .build()
            setAudioAttributes(audioAttributes, true)
            setHandleAudioBecomingNoisy(true)
            setWakeMode(C.WAKE_MODE_LOCAL)
            addListener(this@AudioNewsService)
        }
    }

    private fun initMediaSession() {
        mediaSessionCompat.apply {
            setCallback(mediaSessionCallback)
            isActive = true
        }

        sessionToken = mediaSessionCompat.sessionToken

        mediaSessionConnector.apply {
            setQueueNavigator(object : TimelineQueueNavigator(mediaSessionCompat) {
                override fun getMediaDescription(player: Player, windowIndex: Int) =
                    audioNewsList[windowIndex].getMediaDescription()
            })
            setPlayer(exoPlayer)
        }
    }

    private fun setMediaPlaybackState(state: Int, pos: Int) {
        val playbackStateBuilder = PlaybackStateCompat.Builder()
            .setActiveQueueItemId(pos.toLong())

        if (state == PlaybackStateCompat.STATE_PLAYING || state == PlaybackStateCompat.STATE_BUFFERING ||
            state == PlaybackStateCompat.STATE_SKIPPING_TO_PREVIOUS || state == PlaybackStateCompat.STATE_SKIPPING_TO_NEXT
        ) {
            playbackStateBuilder.setActions(
                PlaybackStateCompat.ACTION_PLAY_PAUSE or PlaybackStateCompat.ACTION_PAUSE or
                        PlaybackStateCompat.ACTION_SKIP_TO_NEXT or PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS or PlaybackStateCompat.ACTION_STOP
            )
        } else {
            playbackStateBuilder.setActions(
                PlaybackStateCompat.ACTION_PLAY_PAUSE or
                        PlaybackStateCompat.ACTION_PLAY or PlaybackStateCompat.ACTION_SKIP_TO_NEXT or
                        PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS or PlaybackStateCompat.ACTION_STOP
            )
        }

        if (state == PlaybackStateCompat.STATE_PAUSED || state == PlaybackStateCompat.STATE_PLAYING) {
            try {
                playbackStateBuilder.setState(state, exoPlayer.currentPosition, 0f)
            } catch (e: IllegalStateException) {
                e.printStackTrace()
                toast(e.message ?: "IllegalStateException")
            }
        } else
            playbackStateBuilder.setState(state, PlaybackStateCompat.PLAYBACK_POSITION_UNKNOWN, 0f)

        mediaSessionCompat.setPlaybackState(playbackStateBuilder.build())
    }

    private fun setCurrentAudio(pos: Int) {
        if (pos >= audioNewsList.size || pos < 0)
            return
        mediaSessionCompat.setMetadata(audioNewsList[pos].buildMetaData())
    }

    private inner class MediaSessionCallback : MediaSessionCompat.Callback() {
        override fun onPlay() {
            super.onPlay()
            exoPlayer.playWhenReady = true
            mediaSessionCompat.isActive = true

            val pos = exoPlayer.currentWindowIndex
            setCurrentAudio(pos)
            setMediaPlaybackState(PlaybackStateCompat.STATE_PLAYING, pos)
        }

        override fun onPause() {
            super.onPause()
            exoPlayer.playWhenReady = false
            setMediaPlaybackState(PlaybackStateCompat.STATE_PAUSED, exoPlayer.currentWindowIndex)
        }

        override fun onSkipToNext() {
            super.onSkipToNext()
            if (exoPlayer.hasNext())
                exoPlayer.next()
        }

        override fun onSkipToPrevious() {
            super.onSkipToPrevious()
            if (exoPlayer.hasPrevious())
                exoPlayer.previous()
        }
    }

    private inner class AudioNewsHandler : Handler.Callback {
        override fun handleMessage(message: Message): Boolean {
            Log.d(TAG, "handle message: ${message.what}")

            when (message.what) {
                MSG_INIT_NEWS -> {
                    audioNewsHandler.removeCallbacksAndMessages(null)
                    audioNewsList.clear()
                    concatenatingMediaSource.clear()

                    launch {
                        loadNews()
                    }
                }

                MSG_STOP_SERVICE -> {
                    audioNewsHandler.removeCallbacksAndMessages(null)
                    stopForeground(true)
                    stopSelf()
                }

                MSG_UPDATE_SOURCE -> {
                    audioNewsHandler.removeMessages(MSG_UPDATE_SOURCE)
                    exoPlayer.prepare(concatenatingMediaSource, false, false)
                }

                MSG_SHOW_NOTI -> {
                    val icon = message.arg1
                    val category = message.obj as String
                    val playWhenReady = message.arg2 == 1
                    val metadataCompat = mediaSessionCompat.controller.metadata
                    launch {
                        val notificationBuilder = buildNotification(
                            icon,
                            category,
                            metadataCompat,
                            mediaSessionCompat.sessionToken
                        )
                        if (playWhenReady)
                            startForeground(MEDIA_NOTI_ID, notificationBuilder.build())
                        else {
                            getSystemService<NotificationManager>()?.notify(
                                MEDIA_NOTI_ID,
                                notificationBuilder.build()
                            )
                            stopForeground(false)
                        }
                    }
                }
            }

            return false
        }
    }
}